File: software/examples/javascript/ExampleCallback.js
var Tinkerforge = require('tinkerforge');

var HOST = 'localhost';
var PORT = 4223;
var UID = 'XYZ'; // Change XYZ to the UID of your DC Bricklet 2.0

var ipcon = new Tinkerforge.IPConnection(); // Create IP connection
var dc = new Tinkerforge.BrickletDCV2(UID, ipcon); // Create device object

ipcon.connect(HOST, PORT,
    function (error) {
        console.log('Error: ' + error);
    }
); // Connect to brickd
// Don't use device before ipcon is connected

ipcon.on(Tinkerforge.IPConnection.CALLBACK_CONNECTED,
    function (connectReason) {
        // The acceleration has to be smaller or equal to the maximum
        // acceleration of the DC motor, otherwise the velocity reached
        // callback will be called too early
        dc.setMotion(4096,
                     16384); // Slow acceleration (12.5 %/s), fast decceleration (50 %/s) for stopping
        dc.setVelocity(32767); // Full speed forward (100 %)

        // Enable motor power
        dc.setEnabled(true);
    }
);

// Register velocity reached callback
dc.on(Tinkerforge.BrickletDCV2.CALLBACK_VELOCITY_REACHED,
    // Use velocity reached callback to swing back and forth
    // between full speed forward and full speed backward
    function (velocity) {
        if(velocity == 32767) {
            console.log('Velocity: Full speed forward, now turning backward');
            dc.setVelocity(-32767);
        }
        else if(velocity === -32767) {
            console.log('Velocity: Full speed backward, now turning forward');
            dc.setVelocity(32767);
        }
        else {
            console.log('Error'); // Can only happen if another program sets velocity
        }
    }
);

console.log('Press key to exit');
process.stdin.on('data',
    function (data) {
        dc.setVelocity(0); // Stop motor before disabling motor power

        setTimeout(function () {
            dc.setEnabled(false); // Disable motor power

            ipcon.disconnect();
            process.exit(0);
        }, 2000); // Wait for motor to actually stop: velocity (100 %) / decceleration (50 %/s) = 2 s
    }
);

File: software/examples/perl/example_callback.pl
#!/usr/bin/perl

# FIXME: This example is incomplete

use strict;
use Tinkerforge::IPConnection;
use Tinkerforge::BrickletDCV2;

use constant HOST => 'localhost';
use constant PORT => 4223;
use constant UID => 'XYZ'; # Change XYZ to the UID of your DC Bricklet 2.0

my $ipcon = Tinkerforge::IPConnection->new(); # Create IP connection
our $dc = Tinkerforge::BrickletDCV2->new(&UID, $ipcon); # Create device object

# Use velocity reached callback to swing back and forth
# between full speed forward and full speed backward
sub cb_velocity_reached
{
    my ($velocity) = @_;

    if ($velocity == 32767)
    {
        print "Velocity: Full speed forward, now turning backward\n";
        $dc->set_velocity(-32767);
    }   
    elsif ($velocity == -32767)
    {
        print "Velocity: Full speed backward, now turning forward\n";
        $dc->set_velocity(32767);
    }    
    else
    {
        print "Error\n"; # Can only happen if another program sets velocity
    }
}

$ipcon->connect(&HOST, &PORT); # Connect to brickd
# Don't use device before ipcon is connected

# The acceleration has to be smaller or equal to the maximum
# acceleration of the DC motor, otherwise the velocity reached
# callback will be called too early
$dc->set_motion(4096,
                16384); # Slow acceleration (12.5 %/s), fast decceleration (50 %/s) for stopping
$dc->set_velocity(32767); # Full speed forward (100 %)

# Register velocity reached callback to subroutine cb_velocity_reached
$dc->register_callback($dc->CALLBACK_VELOCITY_REACHED, 'cb_velocity_reached');

# Enable motor power
$dc->set_enabled(1);

print "Press key to exit\n";
<STDIN>;

$dc->set_velocity(0); # Stop motor before disabling motor power
sleep(2); # Wait for motor to actually stop: velocity (100 %) / decceleration (50 %/s) = 2 s
$dc->set_enabled(0); # Disable motor power

$ipcon->disconnect();
